src/library_manager/lib_manager.c

```diff
@@ -0,0 +1,671 @@
+
+	ret_tmp = sys_mm_drv_unmap_region(va_base_text, st_text_size);
+	if (ret_tmp < 0)
+		ret = -ENOMEM;
+
+	dcache_invalidate_region(va_base_text, st_text_size);
+
+	ret_tmp = sys_mm_drv_unmap_region(va_base_rodata, st_rodata_size);
+	if (ret_tmp < 0)
+		ret = -ENOMEM;
+
+	dcache_invalidate_region(va_base_rodata, st_rodata_size);
+
+	/* There are modules marked as lib_code. This is code shared between several modules inside
+	 * the library. Unload all lib_code modules with last none lib_code module unload.
+	 */
+	if (!mod->type.lib_code && ext_lib->mods_exec_load_cnt > 0)
+		ext_lib->mods_exec_load_cnt--;
+
+	if (ext_lib->mods_exec_load_cnt == 0) {
+		struct sof_man_module *module_entry =
+		    (struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(0));
+		for (size_t idx = 0; idx < desc->header.num_module_entries;
+				++idx, ++module_entry) {
+			if (module_entry->type.lib_code) {
+				ret_tmp =
+				lib_manager_unload_module(lib_id << LIB_MANAGER_LIB_ID_SHIFT |
+							  idx, mod, desc);
+				if (ret_tmp < 0)
+					ret = ret_tmp;
+			}
+		}
+	}
+
+	return ret;
+}
+
+static uint8_t *lib_manager_get_instance_bss_address(uint32_t module_id, uint32_t instance_id,
+						     struct sof_man_module *mod)
+{
+	uint32_t instance_bss_size =
+		 mod->segment[SOF_MAN_SEGMENT_BSS].flags.r.length / mod->instance_max_count;
+	uint32_t inst_offset = instance_bss_size * CONFIG_MM_DRV_PAGE_SIZE * instance_id;
+	uint8_t *va_base =
+		(uint8_t *)(mod->segment[SOF_MAN_SEGMENT_BSS].v_base_addr + inst_offset);
+
+	tr_dbg(&lib_manager_tr,
+	       "lib_manager_get_instance_bss_address() instance_bss_size: 0x%x, pointer: 0x%x",
+	       instance_bss_size, va_base);
+	return va_base;
+}
+
+static int lib_manager_allocate_module_instance(uint32_t module_id, uint32_t instance_id,
+						uint32_t is_pages, struct sof_man_module *mod)
+{
+	int ret;
+	uint32_t bss_size =
+			(mod->segment[SOF_MAN_SEGMENT_BSS].flags.r.length / mod->instance_max_count)
+			 * CONFIG_MM_DRV_PAGE_SIZE;
+	uint8_t *va_base = lib_manager_get_instance_bss_address(module_id, instance_id, mod);
+
+	if ((is_pages * CONFIG_MM_DRV_PAGE_SIZE) > bss_size) {
+		tr_err(&lib_manager_tr,  "is_pages (%d) invalid, required: %d",
+		       is_pages, bss_size / CONFIG_MM_DRV_PAGE_SIZE);
+		return -ENOMEM;
+	}
+	/*
+	 * Map bss memory and clear it.
+	 */
+	ret = sys_mm_drv_map_region(va_base, POINTER_TO_UINT(NULL),
+				    bss_size, SYS_MM_MEM_PERM_RW);
+	if (ret < 0)
+		return -ENOMEM;
+	memset(va_base, 0, bss_size);
+	return 0;
+}
+
+static int lib_manager_free_module_instance(uint32_t module_id, uint32_t instance_id,
+					    struct sof_man_module *mod)
+{
+	int ret;
+	uint32_t bss_size =
+			(mod->segment[SOF_MAN_SEGMENT_BSS].flags.r.length / mod->instance_max_count)
+			 * CONFIG_MM_DRV_PAGE_SIZE;
+	uint8_t *va_base = lib_manager_get_instance_bss_address(module_id, instance_id, mod);
+	/*
+	 * Unmap bss memory and clear it.
+	 */
+	ret = sys_mm_drv_unmap_region(va_base, bss_size);
+	return ret;
```

Comment:
you can directly `return` here. What does `unamp bss memory and clear it`? Does the unamp do the clear also?

Reply:
Does not. Fixed and comment.